File: 02-04-implementing-binary-search/src/main.py
import math
from typing import Any, Sequence


def binary_search(haystack: Sequence, needle: Any) -> bool:
    min_index = 0
    max_index = len(haystack)
    result = False

    while min_index < max_index:
        mid_index = math.floor(min_index + (max_index - min_index) / 2)
        mid_value = haystack[mid_index]

        if mid_value == needle:
            result = True
            break
        elif mid_value < needle:
            min_index = mid_index + 1
        else:
            max_index = mid_index

    return result


if __name__ == "__main__":
    xs = [1, 2, 6, 8, 234, 24524, 123157]
    search_values = [6, 12, 24524]

    for value in search_values:
        result = binary_search(xs, value)
        print(f"is {value} in {xs}: {result}")

File: 02-04-implementing-binary-search/src/main.rs
fn binary_search<T: Ord>(haystack: Vec<T>, needle: T) -> bool {
    use std::cmp::Ordering;

    let mut result = false;
    let mut min_index = 0;
    let mut max_index = haystack.len();

    while min_index < max_index {
        let mid_index = min_index + (max_index - min_index) / 2;
        let mid_value = &haystack[mid_index];

        match mid_value.cmp(&needle) {
            Ordering::Equal => {
                result = true;
                break;
            }
            Ordering::Less => min_index = mid_index + 1,
            _ => max_index = mid_index,
        }
    }

    result
}

fn main() {
    let xs = [1, 2, 6, 8, 234, 24524, 123157];
    let search_values = [6, 12, 24524];

    for value in search_values {
        let result = binary_search(xs.into(), value);
        println!("is {value} in {xs:?}: {result}");
    }
}
